lib/kredis/script_evaluator.rb

```diff
@@ -0,0 +1,17 @@
+module Kredis::ScriptEvaluator
+  class UnfrozenScriptError < StandardError; end
+
+  SCRIPT_DIGESTS = Hash.new do |hash, script|
+    raise UnfrozenScriptError unless script.frozen?
+
+    hash[script] = -Digest::SHA1.hexdigest(script)
+  end
+
+  extend self
+
+  def eval(redis, script, **kwargs)
+    redis.evalsha(SCRIPT_DIGESTS[script], **kwargs)
+  rescue Redis::CommandError
+    redis.eval(script, **kwargs)
```

Comment:
Maybe? I've seen that pattern used elsewhere, but I'm not sure why it would be preferable to just falling back to `eval`.

`eval` also loads the script into the script cache. In the case where you're evaluating a script that is not already in the cache, this code will result in 2 calls to Redis, `evalsha`, and `eval`.

In the `script load` variant, you would wind up with 3 calls to Redis: `evalsha`, `script load`, `evalsha`.

If all else is equal, it seems like you would want the one with less commands. But on the flip side, I see the `script load` solution used a lot in the wild, so maybe that's better for some reason unknown to me.